mmcv/ops/csrc/pytorch/cuda/nattenav.cu

```diff
@@ -0,0 +1,1388 @@
+          nattena_cuda_backward_kernel_fp32_7x7_9x9_32<
+              TILE_9, KTILE_9, KERNEL_SIZE_9, NEIGHBORHOOD_SIZE_9, XYTHREADS_9,
+              scalar_t><<<attn_blocks, attn_threads, 0, stream>>>(
+              d_out_a, d_attn_a, value_a, height, width, batch_size, heads);
+          nattenv_cuda_backward_kernel_fp32<KERNEL_SIZE_9, NEIGHBORHOOD_SIZE_9,
+                                            scalar_t>
+              <<<grid_value, block, 0, stream>>>(d_out_a, d_value_a, attn_a,
+                                                 height, width, heads, dim,
+                                                 n_value);
+        } else if (kernel_size == KERNEL_SIZE_5) {
+          nattena_cuda_backward_kernel_fp32_5x5_32<scalar_t>
+              <<<attn_blocks, attn_threads, 0, stream>>>(
+                  d_out_a, d_attn_a, value_a, height, width, batch_size, heads);
+          nattenv_cuda_backward_kernel_fp32<KERNEL_SIZE_5, NEIGHBORHOOD_SIZE_5,
+                                            scalar_t>
+              <<<grid_value, block, 0, stream>>>(d_out_a, d_value_a, attn_a,
+                                                 height, width, heads, dim,
+                                                 n_value);
+        } else if (kernel_size == KERNEL_SIZE_11) {
+          nattena_cuda_backward_kernel_fp32_11x11_13x13_32<
+              TILE_11_X, TILE_11_Y, KTILE_11_X, KTILE_11_Y, KERNEL_SIZE_11,
+              NEIGHBORHOOD_SIZE_11, XTHREADS_11, YTHREADS_11, scalar_t,
+              scalar_t><<<attn_blocks, attn_threads, 0, stream>>>(
+              d_out_a, d_attn_a, value_a, height, width, batch_size, heads);
+          nattenv_cuda_backward_kernel_fp32<KERNEL_SIZE_11,
+                                            NEIGHBORHOOD_SIZE_11, scalar_t>
+              <<<grid_value, block, 0, stream>>>(d_out_a, d_value_a, attn_a,
+                                                 height, width, heads, dim,
+                                                 n_value);
+        } else if (kernel_size == KERNEL_SIZE_13) {
+          nattena_cuda_backward_kernel_fp32_11x11_13x13_32<
+              TILE_13_X, TILE_13_Y, KTILE_13_X, KTILE_13_Y, KERNEL_SIZE_13,
+              NEIGHBORHOOD_SIZE_13, XTHREADS_13, YTHREADS_13, scalar_t, float>
+              <<<attn_blocks, attn_threads, 0, stream>>>(
+                  d_out_a, d_attn_a, value_a, height, width, batch_size, heads);
+          nattenv_cuda_backward_kernel_fp32<KERNEL_SIZE_13,
+                                            NEIGHBORHOOD_SIZE_13, scalar_t>
+              <<<grid_value, block, 0, stream>>>(d_out_a, d_value_a, attn_a,
+                                                 height, width, heads, dim,
+                                                 n_value);
+        }
+      }));
+  return {d_attn, d_value};
+}
+
+std::vector<torch::Tensor> nattenav_cuda_backward_fp16_tiled_32(
+    const torch::Tensor &d_out, const torch::Tensor &attn,
+    const torch::Tensor &value) {
```

Comment:
add device guard `at::cuda::CUDAGuard device_guard(value.device());` before kernel launch to avoid error on non-default device.

Reply:
I have added a device guard before every function using kernel launch.